test/celeritas/optical/RoughnessCalculator.test.cc

```diff
@@ -0,0 +1,128 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/RoughnessCalculator.test.cc
+//---------------------------------------------------------------------------//
+#include <memory>
+#include <random>
+
+#include "corecel/math/ArrayOperators.hh"
+#include "corecel/random/DiagnosticRngEngine.hh"
+#include "celeritas/ext/RootFileManager.hh"
+#include "celeritas/optical/surface/calc/GaussianRoughnessCalculator.hh"
+#include "celeritas/optical/surface/calc/SmearRoughnessCalculator.hh"
+
+#include "celeritas_test.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+namespace test
+{
+using namespace ::celeritas::test;
+//---------------------------------------------------------------------------//
+
+class RoughnessCalculatorTest : public ::celeritas::test::Test
+{
+  public:
+    using RandomEngine = DiagnosticRngEngine<std::mt19937>;
+
+    RandomEngine& rng()
+    {
+        rng_.reset_count();
+        return rng_;
+    }
+
+    // Sample a roughness calculator and collect distribution moments
+    template<class C>
+    Array<real_type, 4> sample_moments(C const& sample_normal,
+                                       Real3 const& normal,
+                                       Real3 const& momentum)
+    {
+        EXPECT_TRUE(dot_product(normal, momentum) < 0);
+
+        auto& rng = this->rng();
+
+        size_type num_samples = 1000;
+
+        // 1st moment of cos with normal
+        // 2nd moment of cos with normal
+        // 1st moment of cos with momentum
+        // 2nd moment of cos with momentum
+        Array<real_type, 4> moments{0, 0, 0, 0};
+
+        for ([[maybe_unused]] auto i : range(num_samples))
+        {
+            Real3 local_normal = sample_normal(rng);
+
+            real_type normal_cos = dot_product(normal, local_normal);
+            real_type momentum_cos = dot_product(momentum, local_normal);
+
+            EXPECT_TRUE(is_soft_unit_vector(local_normal));
+            EXPECT_TRUE(momentum_cos < 0);
+
+            moments[0] += normal_cos;
+            moments[1] += ipow<2>(normal_cos);
+            moments[2] += momentum_cos;
+            moments[3] += ipow<2>(momentum_cos);
```

Comment:
Is this the best way of measuring the sampled normal since the incident normal and momentum are basically hard-coded? (With the values chosen, the `x` component isn't even contributing to the sample result.) I think a better test might be to bin the dot product with the surface normal (and you can put the rejection loop in here). Have you seen the `HistogramSampler` helper class? That could help you easily test a couple of different roughness values of the smearing and gaussian to really observe the effects..